core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/udf/PythonUDFOperator.java

```diff
@@ -0,0 +1,700 @@
+        }catch(Exception e){
+            closeClientAndServer();
+            throw new DataflowException(e.getMessage(), e);
+        }
+
+        Schema inputSchema = inputOperator.getOutputSchema();
+
+        // generate output schema by transforming the input schema
+        outputSchema = transformToOutputSchema(inputSchema);
+
+        cursor = OPENED;
+
+        tupleToPythonSchema = convertToArrowSchema(inputSchema);
+        innerIndexMap = new HashMap<>();
+    }
+
+    /**
+     * For every batch, the operator calls {@code flightClient.doAction(new Action("compute"))} to tell the server to
+     * compute sentiments of the specific table that was sent earlier. The server executes computation,
+     * and returns back a success message when computation is finished.
+     * @return Whether the buffer is empty
+     */
+    private boolean computeTupleBuffer() {
+        tupleBuffer = new ArrayList<Tuple>();
+        int i = 0;
+        while (i < predicate.getBatchSize()){
+            Tuple inputTuple;
+            if ((inputTuple = inputOperator.getNextTuple()) != null) {
+                tupleBuffer.add(inputTuple);
+                i++;
+            } else {
+                break;
+            }
+        }
+        if (tupleBuffer.isEmpty()) {
+            return false;
+        }
+        writeArrowStream(tupleBuffer, globalRootAllocator, tupleToPythonSchema, "toPython");
+        return true;
+    }
+
+    @Override
+    public Tuple getNextTuple() throws TexeraException {
+        if (cursor == CLOSED) {
+            return null;
+        }
+
+        // Not necessarily one input to one output mapping now.
+
+        if (resultQueue != null) {
+            // UDF execution has been made, retrieve results.
+            if (resultQueue.isEmpty()) {
+                // Try to make one more batch execution
+                if (computeTupleBuffer()) {
+                    executeUDF();
+                }
+                // All the tuples have been sent to Python and no result remain.
+                if (resultQueue.isEmpty()) return null;
+                else return resultQueue.remove();
+            } else return resultQueue.remove();
+        } else {
+            // No execution of UDF has been made yet.
+            if (tupleBuffer == null){
+                if (computeTupleBuffer()) {
+                    executeUDF();
+                } else {
+                    // No data at all
+                    tupleBuffer = null;
+                    return null;
+                }
+            }
+            if (resultQueue == null || resultQueue.isEmpty()) return null;
+            else return resultQueue.remove();
+        }
+
+    }
+
+    // Process the data file using NLTK
+    private void executeUDF() {
+        try{
+//            System.out.println("Flight Client:\t" + new String(
+            flightClient.doAction(new Action("compute")).next().getBody();
```

Comment:
I find we are not checking the return value in many places. Since this PR will is a more formal implemenation, rather than being a prototype, we need to formally define the interfaces.

Reply:
What do you mean by return value? Can you show where to change?